### ISSUE TITLE
dCache does not honour the kXR_mkpath option

### ISSUE THREAD

**Issue by paulmillar** [dCache/dcache#4178]
The kXR_open and kXR_mkdir commands accept the kXR_mkpath optional flag, which tells the server to create parent path elements if they do not already exit.  This is similar to the mkdir(1) `-p` option in Unix shells.

dCache does not support this flag; in fact, it completely ignores it.  Note that similar support already exists for WebDAV and SRM, so this should be relatively easy to fix.

Clients, such as rucio, currently require this functionality.

**Comment by mlassnig** [comment]
@mlassnig adding as watcher

**Comment by paulmillar** [comment]
Example demonstrating the problem:

```
gfal-mkdir root://prometheus.desy.de:1095//VOs/dteam/tpc-test/test/a5/b3
gfal-mkdir error: 5 (Input/output error) - Failed to create directory
root://prometheus.desy.de:1095//VOs/dteam/tpc-test/test/a5/b3
(Input/output error)
```

Example showing that creating each directory (in turn) works as expected (assuming `/VOs/dteam/tpc-test` already exists)
```
gfal-mkdir root://prometheus.desy.de:1095//VOs/dteam/tpc-test/test
gfal-mkdir root://prometheus.desy.de:1095//VOs/dteam/tpc-test/test/a5
gfal-mkdir root://prometheus.desy.de:1095//VOs/dteam/tpc-test/test/a5/b3
```

**Comment by paulmillar** [comment]
Links to the relevant sections from the [Xrootd v3.1.0 specification](http://xrootd.org/doc/dev45/XRdv310.htm): [kXR_mkdir](http://xrootd.org/doc/dev45/XRdv310.htm#_Toc464248821), [kXR_open](http://xrootd.org/doc/dev45/XRdv310.htm#_Toc464248819).

**Comment by alrossi** [comment]
To say that dCache does not support this is inaccurate.  The code is there in the protocol.  See:

org.dcache.xrootd.protocol.messages.OpenRequest
org.dcache.xrootd.protocol.messages.MkDirRequest

The standalone server implements mkdir -p as it should.

Similarly, the door code processes the parameter:

 _door.createDirectory(createFullPath(req.getPath()),
                                  req.shouldMkPath(),
                                  req.getSubject(),
                                  _authz);

and this in turn calls the pnfsHandler accordingly:

    if (createParents) {
            pnfsHandler.createDirectories(path);
    } else {
            pnfsHandler.createPnfsDirectory(path.toString());
    }

So there is something deeper that is wrong here.

**Comment by alrossi** [comment]
Also in the Transfer write code (from the doOnOpen entry point)

                if (createDir) {
                    transfer.createNameSpaceEntryWithParents();
                } else {
                    transfer.createNameSpaceEntry();
                }

**Comment by paulmillar** [comment]
OK, me bad.  It *should* work but is broken.

Here's a simple test demonstrating it currently doesn't work with xrdfs (as I don't have gfal-mkdir installed on my desktop)

```
paul@celebrimbor:~/git/dCache (master)$ xrdfs root://prometheus.desy.de:1095 mkdir -p /Users/paul/Private/test-2/test-3
[ERROR] Server responded with an error: [3005] No such file or directory: /Users/paul/Private/test-2/test-3

paul@celebrimbor:~/git/dCache (master)$ xrdfs root://prometheus.desy.de:1095 mkdir -p /Users/paul/Private/test-2
paul@celebrimbor:~/git/dCache (master)$ xrdfs root://prometheus.desy.de:1095 mkdir -p /Users/paul/Private/test-2/test-3
paul@celebrimbor:~/git/dCache (master)$ 
```

**Comment by mksahakyan** [comment]
what iI can see here is that    

 mkdir -p  option=1, 1&256! = 256
 mkdir   option=0,   0&256! = 256

so 
    public boolean shouldMkPath() {
        return (this.getOptions() & 256) == 256;
    }
returns false.

**Comment by paulmillar** [comment]
This is interesting.  The xrootd spec fails to mention the value of `kXR_mkpath`, however, the xrootd code defines the value as 256:

https://github.com/xrootd/xrootd/blob/master/src/XProtocol/XProtocol.hh#L222

Yet the xrootd client seems to set bit-0 (i.e. 1) rather than bit-8 (i.e., 256).

The options field is 16-bit (i.e., two bytes), so could this be an endian issue?

**Comment by alrossi** [comment]
I am not sure this would be a byte order issue.   We don't seem to encounter that elsewhere.

Is there some way to test the kXR_mkpath flag when doing a file write instead of using xrdfs mkdir?  The latter seems to have a different code path, and it looks like the value of the flag there may be different.  

int XrdPosixXrootd::Mkdir(const char *path, mode_t mode)
{
  XrdPosixAdmin admin(path);
  XrdCl::MkDirFlags::Flags flags;

// Preferentially make the whole path unless told otherwise
//
   flags = (mode & S_ISUID ? XrdCl::MkDirFlags::None
                           : XrdCl::MkDirFlags::MakePath);

// Make sure the admin is OK
//
  if (!admin.isOK()) return -1;

// Issue the mkdir
//
   return XrdPosixMap::Result(admin.Xrd.MkDir(admin.Url.GetPathWithParams(),
                                              flags,
                                              XrdPosixMap::Mode2Access(mode))
                                             );
}

MakePath = 1, None = 0

This is different from the POSIX open flags, which match the above:

if (oflags & O_CREAT)
      {XOflags |= (oflags & O_EXCL ? XrdCl::OpenFlags::New
                                   : XrdCl::OpenFlags::Delete);
       XOflags |= XrdCl::OpenFlags::MakePath;
       XOmode   = XrdPosixMap::Mode2Access(mode);
      }
      else if (oflags & O_TRUNC && Opts & XrdPosixFile::isUpdt)
              XOflags |= XrdCl::OpenFlags::Delete;

here MakePath is set to :  

      MakePath = kXR_mkpath,     //!< Create directory path if it does not
                                 //!< already exist

and there the value is 256.

So our server may need to do something different on mkdir from what it does on open for write. ???

**Comment by alrossi** [comment]
Indeed.  This is either a bug, or bad protocol definition.

the client's Open enum:

struct OpenFlags
  {
    //--------------------------------------------------------------------------
    //! Open flags, may be or'd when appropriate
    //--------------------------------------------------------------------------
    enum Flags
    {
      None     = 0,              //!< Nothing
      Compress = kXR_compress,   //!< Read compressed data for open (ignored),
                                 //!< for kXR_locate return unique hosts
      Delete   = kXR_delete,     //!< Open a new file, deleting any existing
                                 //!< file
      Force    = kXR_force,      //!< Ignore file usage rules, for kXR_locate
                                 //!< it means ignoreing network dependencies
      MakePath = kXR_mkpath,     //!< Create directory path if it does not
                                 //!< already exist
      New      = kXR_new,        //!< Open the file only if it does not already
                                 //!< exist
      NoWait   = kXR_nowait,     //!< Open the file only if it does not cause
                                 //!< a wait. For locate: provide a location as
                                 //!< soon as one becomes known. This means
                                 //!< that not all locations are necessarily
                                 //!< returned. If the file does not exist a
                                 //!< wait is still imposed.
      Append   = kXR_open_apnd,  //!< Open only for appending
      Read     = kXR_open_read,  //!< Open only for reading
      Update   = kXR_open_updt,  //!< Open for reading and writing
      Write    = kXR_open_wrto,  //!< Open only for writing
      POSC     = kXR_posc,       //!< Enable Persist On Successful Close
                                 //!< processing
      Refresh  = kXR_refresh,    //!< Refresh the cached information on file's
                                 //!< location. Voids NoWait.
      Replica  = kXR_replica,    //!< The file is being opened for replica
                                 //!< creation
      SeqIO    = kXR_seqio,      //!< File will be read or written sequentially
      PrefName = kXR_prefname    //!< Hostname response is prefered, applies
                                 //!< only to FileSystem::Locate
    };
  };
  XRDOUC_ENUM_OPERATORS( OpenFlags::Flags )

the client's Mkdir enum:

struct MkDirFlags
  {
    enum Flags
    {
      None     = 0,  //!< Nothing special
      MakePath = 1   //!< create the entire directory tree if it doesn't exist
    };
  };
  XRDOUC_ENUM_OPERATORS( MkDirFlags::Flags )

**Comment by paulmillar** [comment]
OK, the story so far.

The xrootd protocol spec (v3.1.0) describes how `kXR_mkdir` and `kXR_open` operations both have an `options` field, which act as a bit mask for flags that affect the behaviour of these commands.  According to the spec, both accept `kXR_mkpath` as one of these flags, but fails to define this.

The xrootd code parses these values, but this make-parent-values has different values for `kXR_mkdir` and `kXR_open` (1 and 256 respectively).

The ironic part is that xrootd4j *does* define the correct value for the `kXR_mkdir` options field, but as `kXR_mkdirpath` rather than `kXR_mkpath`.

So, in summary, the xrootd protocol spec is incomplete (lacking definitions) and wrong (says same constant is used for kXR_mkdir kXR_open).

I've opened an issue describing this problem:

https://github.com/xrootd/xrootd/issues/815

The good news is it should be trivial to fix.

**Comment by alrossi** [comment]
Good. Thanks for confirming this.   What I was thinking as well.

By the way, a lot of the constants defined in the protocol.hh are not specified in the protocol document.  You have to look at the C++ header to make sure we have the correct values in our Java implementation.
